pulser/waveforms.py

```diff
@@ -794,6 +794,74 @@ def __init__(
             self._area / float(np.sum(self._norm_samples)) / 1e-3
         )
 
+    @classmethod
+    @parametrize
+    def from_max_val(
+        cls,
+        max_val: Union[float, Parametrized],
+        area: Union[float, Parametrized],
+        beta: Optional[Union[float, Parametrized]] = 14.0,
+    ) -> KaiserWaveform:
+        """Creates a Kaiser waveform with a threshold on the maximum value.
+
+        Instead of defining a duration, the waveform is defined by its area and
+        the maximum value. The duration is chosen so that the maximum value is
+        not surpassed, but approached as closely as possible.
+
+        Args:
+            max_val (float): The maximum value threshold (in rad/µs). If
+                negative, it is taken as the lower bound i.e. the minimum
+                value that can be reached. The sign of `max_val` must match the
+                sign of `area`.
+            area (float): The area under the waveform.
+            beta (Optional[float]): The beta parameter of the Kaiser window.
+                The default value is 14.
+        """
+        # Compute the ratio area / duration for a long duration
+        # and use this value for a first guess of the best duration
+
+        max_val = cast(float, max_val)
+        area = cast(float, area)
+
+        ratio: float = max_val * np.sum(np.kaiser(100, beta)) / 100
+        duration_guess: int = int(area * 1000.0 / ratio)
+
+        duration_best: int = 0
+
+        if duration_guess < 11:
+            # Because of the seesawing effect on short durations,
+            # all solutions must be tested to find the best one
+
+            max_val_best: float = 0
+            for duration in range(1, 16):
+                kaiser_temp = np.kaiser(duration, beta)
+                scaling_temp = 1000 * area / np.sum(kaiser_temp)
+                max_val_temp = np.max(kaiser_temp) * scaling_temp
+                if max_val_best < max_val_temp <= max_val:
+                    max_val_best = max_val_temp
+                    duration_best = duration
+
+        else:
+            # Increase or decrease duration depending on
+            # the max value for the guessed duration
+
+            kaiser_guess = np.kaiser(duration_guess, beta)
+            scaling_guess = 1000 * area / np.sum(kaiser_guess)
+
+            step = 1 if np.max(kaiser_guess) * scaling_guess > max_val else -1
```

Comment:
Just so I'm sure I have this right: you have that `step = np.sign(np.max(kaiser_guess) * scaling_guess - max_val)` when they are not equal, right?

Reply:
`step` tells us if the duration must be incremented (the guess is greater than `max_val`) or decremented (the guess is less than `max_val`).

So, yes, both lines of code are more or less equivalent.

The only difference is that I never have `step == 0` that could end up in a infinite loop.

But something might be wrong here... I didn't take into account a perfect match before looping...

Reply:
I think you should have this check for `np.max(kaiser_guess) * scaling_guess >= max_val` instead. That way, when they match perfectly, it will set `step == 1` which means it won't make the adjustment of `duration_best` at the end. Combine this with it not entering the while loop if you set the `max_val_temp` as I suggested before, and you go straight to the right result.